File: algorithms/2_medium/M/03952_maximum_total_value_of_covered_indices/dp_o1_space.c
static inline int64_t max_int64(const int64_t a, const int64_t b)
{
    return a > b ? a : b;
}

long long maxTotal(const int* nums, const int len, const char* s) {
    assert(len > 0 && nums != NULL && s != NULL && (size_t)len == strlen(s));
    int64_t prev_prev = 0;
    int64_t prev = 0;
    if (s[0] == '1') {
        prev = nums[0];
    }

    for (int i = 1; i < len; i += 1) {
        if (s[i] == '1') {
            const int64_t curr = max_int64(prev_prev + nums[i - 1], prev + nums[i]);
            prev = max_int64(prev, prev_prev + nums[i - 1]);

            prev_prev = prev;
            prev = curr;
        } else {
            prev_prev = prev;
        }
    }

    return prev;
}

File: algorithms/2_medium/M/03952_maximum_total_value_of_covered_indices/dp_o1_space.rs
impl Solution {
    pub fn max_total(nums: Vec<i32>, s: String) -> i64 {
        let len = nums.len();
        let bytes = s.as_bytes();
        assert!(len == bytes.len() && len > 0);
        let (mut prev_prev, mut prev) = (0_i64, 0_i64);
        if bytes[0] == b'1' {
            prev = nums[0] as i64;
        }

        for i in 1..len {
            if bytes[i] == b'1' {
                let curr = (prev_prev + nums[i - 1] as i64).max(prev + nums[i] as i64);
                let new_prev = prev.max(prev_prev + nums[i - 1] as i64);
                
                (prev_prev, prev) = (new_prev, curr);
            } else {
                prev_prev = prev;
            };
        }

        prev
    }
}
